node_modules/col-assets/lib/api.js

```diff
@@ -222,22 +228,118 @@ var createLink = module.exports.createLink = function(ctx, title, url, opts, cal
+ * @param  {String}         title                           The title of the file
+ * @param  {String}         file                            The file to create
+ * @param  {Object}         [opts]                          A set of optional parameters
+ * @param  {Number[]}       [opts.categories]               The ids of the categories to which the file should be associated
+ * @param  {String}         [opts.description]              The description of the file
+ * @param  {Function}       callback                        Standard callback function
+ * @param  {Object}         callback.err                    An error that occurred, if any
+ * @param  {Asset}          callback.asset                  The created link asset
+ */
+var createFile = module.exports.createFile = function(ctx, title, file, opts, callback) {
+  file = file || {};
+  // Default the title to the provided filename
+  title = title || file.filename;
+
+  // Default the optional values
+  opts = opts || {};
+  opts.categories = opts.categories || [];
+
+  // Parameter validation
+  var validationSchema = Joi.object().keys({
+    'title': Joi.string().required(),
+    'file': Joi.object().required().keys({
+      'field': Joi.any().valid('file'),
+      'encoding': Joi.string().required(),
+      'mimetype': Joi.string().required(),
+      'truncated': Joi.boolean().valid(false),
+      'uuid': Joi.string().guid().required(),
+      'file': Joi.string().required(),
+      'filename': Joi.string().required()
+    }),
+    'opts': Joi.object().keys({
+      'categories': Joi.array().unique().items(Joi.number()).optional(),
+      'description': Joi.string().allow('').optional(),
+      'source': Joi.string().uri().optional()
+    })
+  });
+
+  var validationResult = Joi.validate({
+    'title': title,
+    'file': file,
+    'opts': opts
+  }, validationSchema);
+
+  if (validationResult.error) {
+    return callback({'code': 400, 'msg': validationResult.error.details[0].message});
+  }
+
+  // For security reasons, the canvas object on the course API only contains the `use_https` and
+  // `canvas_api_domain` attributes. By reloading the instance all other attributes such as the
+  // `canvas_api_key` will be retrieved
+  ctx.course.canvas.reload().complete(function(err) {
     if (err) {
       return callback(err);
     }
 
-    // Create the link asset in the DB
-    var asset = {
-      'course_id': ctx.course.id,
-      'user_id': ctx.user.id,
-      'type': 'link',
-      'title': title,
-      'url': url,
-      'description': opts.description,
-      'source': opts.source
-    };
+    // Upload the file to Canvas
+    uploadFileToCanvas(ctx, file.file, function(err, canvasItem) {
+      if (err) {
+        return callback(err);
+      }
+
+      // Create the file asset in the DB
+      var asset = {
+        'course_id': ctx.course.id,
+        'user_id': ctx.user.id,
+        'type': 'file',
+        'title': title,
+        'download_url': canvasItem.url,
```

Comment:
Can we modify the download URL so it is proxied through the Collabosphere server? Having the download URL point to Canvas directly will cause the Whiteboard tool to not be able to load the images because of cross-origin policy issues.